src/python/pants/engine/internals/platform_rules.py

```diff
@@ -15,5 +24,43 @@ def current_platform(env_tgt: EnvironmentTarget) -> Platform:
     return Platform(env_tgt.val[DockerPlatformField].normalized_value)
 
 
+@rule
+async def complete_environment_vars(
+    session_values: SessionValues, env_tgt: EnvironmentTarget
+) -> CompleteEnvironmentVars:
+    if not env_tgt.val or not env_tgt.val.has_field(DockerImageField):
+        return session_values[CompleteEnvironmentVars]
+    env_process_result = await Get(
+        ProcessResult,
+        Process(
+            ["env"],
```

Comment:
This won't work on Windows and not even on some Linux images. It would be good to have a pre-built static env-like binary for each platform we support (could write in rust), to inject in the container and use.

Reply:
Doesn't even need to be Rust. A small statically-linked C shim would be just fine (and super small binary size).

Reply:
```
#include <stdio.h>

extern char **environ;

int main() {
  for (char** entry = environ ; *entry != NULL; entry++) {
    printf("%s\n", *entry);
  }
  return 0;
}
```

macOS/arm binary comes out at 33,472 bytes. With such a small size, we could just embed the binaries in Pants as resources.

Reply:
And then use `Platform` from https://github.com/pantsbuild/pants/pull/16867 to conditionally load the platform-specific binary, right?

Reply:
And it can be switched to use `\0` as the separator so newlines come through okay.